File: mahaam-api-cs/Src/Feat/Plan/PlanService.cs
using System.Transactions;
using Mahaam.Feat.Users;
using Mahaam.Infra;

namespace Mahaam.Feat.Plans;
public interface IPlanService
{
	Task<Plan> GetOne(Guid planId);
	Task<List<Plan>> GetMany(string type);
	Task<Guid> Create(PlanIn plan);
	Task Update(PlanIn plan);
	Task Delete(Guid id);
	Task Share(Guid id, string email);
	Task Unshare(Guid id, string email);
	Task Leave(Guid id);
	Task UpdateType(Guid id, string type);
	Task ReOrder(string type, int oldOrder, int newOrder);
	Task ValidateUserOwnsThePlan(Guid planId);
}

public class PlanService(IPlanRepo planRepo, IPlanMembersRepo planMembersRepo, IUserRepo userRepo, ISuggestedEmailsRepo suggestedEmailsRepo, ILog log) : IPlanService
{

	public async Task<Plan> GetOne(Guid planId)
	{
		var plan = await planRepo.GetOne(planId);
		if (plan is { IsShared: true }) plan.Members = await planMembersRepo.GetUsers(planId);
		return plan;
	}

	public async Task<List<Plan>> GetMany(string type)
	{
		// plans of the user shared or not
		var plans = await planRepo.GetMany(Req.UserId, type);
		if (Req.IsLoggedIn)
		{
			// plans of others shared with the user
			var sharedPlans = await planMembersRepo.GetOtherPlans(Req.UserId);
			plans.AddRange(sharedPlans);
		}
		return plans;
	}

	public async Task<Guid> Create(PlanIn plan)
	{
		var userId = Req.UserId;
		var count = await planRepo.GetCount(userId, "Main");
		if (count >= 100) throw new LogicException("max_is_100", "Max is 100");

		return await planRepo.Create(plan);
	}

	public async Task Update(PlanIn plan)
	{
		await ValidateUserOwnsThePlan(plan.Id);
		await planRepo.Update(plan);
	}

	public async Task Delete(Guid id)
	{
		await ValidateUserOwnsThePlan(id);
		using var scope = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled);
		await planRepo.RemoveFromOrder(Req.UserId, id);
		await planRepo.Delete(id); // This will delete all related records as it is a cascade delete
		scope.Complete();
	}

	public async Task Share(Guid id, string email)
	{
		await ValidateUserLoggedIn();
		await ValidateUserOwnsThePlan(id);
		var user = await userRepo.GetOne(email) ?? throw new ArgumentException("emailnotfound", $"email:{email} was not found");

		if (user.Id.Equals(Req.UserId))
			throw new LogicException("not_allowed_to_share_with_creator", "Not allowed to share with creator");

		var limit = 20;
		var plan = await planRepo.GetOne(id);
		if (plan is { IsShared: true })
		{
			var membersCount = await planMembersRepo.GetUsersCount(id);
			if (membersCount >= limit) throw new LogicException("max_is_20", "Max is 20");
		}
		else
		{
			var plansCount = await planMembersRepo.GetPlansCount(Req.UserId);
			if (plansCount >= limit) throw new LogicException("max_is_20", "Max is 20");
		}
		await planMembersRepo.Create(id, user.Id);
		await suggestedEmailsRepo.Create(Req.UserId, email);
		var usr = await userRepo.GetOne(Req.UserId);
		await suggestedEmailsRepo.Create(user.Id, usr!.Email!);
	}

	public async Task Unshare(Guid id, string email)
	{
		await ValidateUserLoggedIn();
		await ValidateUserOwnsThePlan(id);
		var user = await userRepo.GetOne(email);
		if (user is null)
		{
			throw new ArgumentException("email_not_found", $"email:{email} was not found");
		}
		await planMembersRepo.Delete(id, user.Id);
	}

	public async Task Leave(Guid id)
	{
		await ValidateUserLoggedIn();
		int deletedRecords = await planMembersRepo.Delete(id, Req.UserId);
		if (deletedRecords == 1)
			log.Info($"user {Req.UserId} left plan {id}");
		else
			throw new ArgumentException($"userId={Req.UserId} unable to leave planId={id}");

	}

	public async Task UpdateType(Guid id, string type)
	{
		await ValidateUserOwnsThePlan(id);
		var count = await planRepo.GetCount(Req.UserId, type);
		if (count >= 100) throw new LogicException("max_is_100", "Max is 100");

		using var scope = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled);
		await planRepo.RemoveFromOrder(Req.UserId, id);
		await planRepo.UpdateType(Req.UserId, id, type);
		scope.Complete();
	}


	public async Task ReOrder(string type, int oldOrder, int newOrder)
	{
		var count = await planRepo.GetCount(Req.UserId, type);
		if (oldOrder > count || newOrder > count)
			throw new InputException("oldOrder and newOrder should be less than " + count);

		await planRepo.UpdateOrder(Req.UserId, type, oldOrder, newOrder);
	}

	public async Task ValidateUserOwnsThePlan(Guid planId)
	{
		var plan = await planRepo.GetOne(planId);
		if (plan is null) throw new ArgumentException("planId not found");
		if (!plan!.User.Id.Equals(Req.UserId))
			throw new UnauthorizedException("User does not own this plan");
	}

	private async Task ValidateUserLoggedIn()
	{
		var user = await userRepo.GetOne(Req.UserId);
		if (user is null || user.Email == null)
		{
			throw new LogicException("you_are_not_logged_in", "You are not logged In");
		}
	}
}

File: mahaam-api-go/feat/service/plan_service.go
package service

import (
	"fmt"
	"mahaam-api/feat/models"
	"mahaam-api/feat/repo"
	"mahaam-api/infra/dbs"

	"github.com/google/uuid"
	"github.com/jmoiron/sqlx"
)

type PlanService interface {
	GetOne(planID UUID) *Plan
	GetMany(userID UUID, planType string) []Plan
	Create(userID UUID, plan PlanIn) UUID
	Update(userID UUID, plan *PlanIn)
	Delete(userID UUID, id UUID)
	Share(userID UUID, id UUID, email string)
	Unshare(userID UUID, id UUID, email string)
	Leave(userID UUID, id UUID)
	UpdateType(userID UUID, id UUID, planType string)
	ReOrder(userID UUID, planType string, oldOrder, newOrder int)
	ValidateUserOwnsThePlan(userID UUID, planID UUID)
}

type planService struct {
	planRepo            repo.PlanRepo
	planMembersRepo     repo.PlanMembersRepo
	userRepo            repo.UserRepo
	suggestedEmailsRepo repo.SuggestedEmailRepo
}

func NewPlanService(db *sqlx.DB, planRepo repo.PlanRepo, planMembersRepo repo.PlanMembersRepo, userRepo repo.UserRepo, suggestedEmailsRepo repo.SuggestedEmailRepo) PlanService {
	return &planService{
		planRepo:            planRepo,
		planMembersRepo:     planMembersRepo,
		userRepo:            userRepo,
		suggestedEmailsRepo: suggestedEmailsRepo,
	}
}

func (s *planService) GetOne(planID UUID) *Plan {
	plan := s.planRepo.GetOne(planID)
	if plan.IsShared {
		users := s.planMembersRepo.GetUsers(planID)
		plan.Members = users
	}
	return plan
}

func (s *planService) GetMany(userID UUID, planType string) []Plan {
	plans := s.planRepo.GetMany(userID, planType)
	sharedPlans := s.planMembersRepo.GetOtherPlans(userID)
	return append(plans, sharedPlans...)
}

func (s *planService) Create(userID UUID, plan PlanIn) UUID {
	count := s.planRepo.GetCount(userID, string(models.PlanTypeMain))
	if count >= 100 {
		panic(models.LogicErr("maximum of 100 plans reached", "max_is_100"))
	}

	var planID UUID
	err := dbs.WithTx(func(tx *sqlx.Tx) error {
		planID = s.planRepo.Create(tx, userID, plan)
		return nil
	})

	if err != nil {
		return uuid.Nil
	}
	return planID
}

func (s *planService) Update(userID UUID, plan *PlanIn) {
	s.ValidateUserOwnsThePlan(userID, plan.ID)
	s.planRepo.Update(plan)
}

func (s *planService) Delete(userID UUID, id UUID) {
	s.ValidateUserOwnsThePlan(userID, id)
	dbs.WithTx(func(tx *sqlx.Tx) error {
		s.planRepo.RemoveFromOrder(tx, userID, id)
		s.planRepo.Delete(tx, id)
		return nil
	})
}

func (s *planService) Share(userID UUID, id UUID, email string) {
	s.ValidateUserOwnsThePlan(userID, id)
	user := s.userRepo.GetOneByEmail(email)
	if user == nil {
		panic(models.NotFoundErr("email not found"))
	}
	if user.ID == userID {
		panic(models.LogicErr("not allowed to share with creator", "not_allowed_to_share_with_creator"))
	}

	const limit = 20
	plan := s.planRepo.GetOne(id)
	if plan.IsShared {
		count := s.planMembersRepo.GetUsersCount(id)
		if count >= limit {
			panic(models.LogicErr("maximum of 20 shares reached", "max_is_20"))
		}
	} else {
		count := s.planMembersRepo.GetPlansCount(userID)
		if count >= limit {
			panic(models.LogicErr("maximum of 20 shares reached", "max_is_20"))
		}
	}

	s.planMembersRepo.Create(id, user.ID)
	// No transaction needed for suggested emails, as it's just a suggestion
	s.suggestedEmailsRepo.Create(userID, email)
	creator := s.userRepo.GetOne(userID)
	if creator.Email != nil {
		s.suggestedEmailsRepo.Create(user.ID, *creator.Email)
	}
}

func (s *planService) Unshare(userID UUID, id UUID, email string) {
	s.ValidateUserOwnsThePlan(userID, id)
	user := s.userRepo.GetOneByEmail(email)
	if user == nil {
		panic(models.NotFoundErr("email not found"))
	}
	s.planMembersRepo.Delete(id, user.ID)
}

// Leave allows a user to leave a shared plan
func (s *planService) Leave(userID UUID, id UUID) {
	rows := s.planMembersRepo.Delete(id, userID)
	if rows != 1 {
		panic(models.LogicErr(fmt.Sprintf("user cannot leave plan: userId=%s, planId=%s", userID, id), "user_cannot_leave_plan"))
	}
}

func (s *planService) UpdateType(userID UUID, id UUID, planType string) {
	s.ValidateUserOwnsThePlan(userID, id)
	count := s.planRepo.GetCount(userID, planType)
	if count >= 100 {
		panic(models.LogicErr("maximum of 100 plans reached", "max_is_100"))
	}

	dbs.WithTx(func(tx *sqlx.Tx) error {
		s.planRepo.RemoveFromOrder(tx, userID, id)
		s.planRepo.UpdateType(tx, userID, id, planType)
		return nil
	})
}

func (s *planService) ReOrder(userID UUID, planType string, oldOrder, newOrder int) {
	count := s.planRepo.GetCount(userID, planType)
	if oldOrder > int(count) || newOrder > int(count) {
		panic(models.InputErr(fmt.Sprintf("oldOrder and newOrder should be less than %d", count)))
	}
	s.planRepo.UpdateOrder(userID, planType, oldOrder, newOrder)
}

func (s *planService) ValidateUserOwnsThePlan(userID UUID, planID UUID) {
	plan := s.planRepo.GetOne(planID)
	if plan == nil {
		panic(models.NotFoundErr("plan not found"))
	}

	if plan.User.ID != userID {
		panic(models.ForbiddenErr("user does not own this plan"))
	}
}
